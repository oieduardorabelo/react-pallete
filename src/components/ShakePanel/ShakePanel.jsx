import React from 'react';

import { stopPropagation } from '../../domain';

import styles from '../ColorColumn/ColorColumnStyles'

const panelStyles = styles.panel();
const panelOverlayStyles = styles.panelOverlay();

const displayFlex = 'flex';
panelStyles.display = displayFlex;
panelOverlayStyles.display = displayFlex;

const ShakePanel = (props) => (
  <div>
    <div style={panelOverlayStyles} onClick={props.onClickToggleShakePanel} />

    <div style={panelStyles} onClick={stopPropagation}>
      <button
        onClick={props.onClickToggleShakePanel}
        style={styles.panelClose()}
        type="button"
      >
        Close
      </button>

      <span>Shake it off!</span>
      <p>When you shake your phone, we set random colors to your pallete!</p>
    </div>
  </div>
)

ShakePanel.propTypes = {
  onClickToggleShakePanel: React.PropTypes.func.isRequired,
}

export default ShakePanel;
