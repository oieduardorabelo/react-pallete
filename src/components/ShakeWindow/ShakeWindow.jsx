import React from 'react';
import ShakeJS from 'shake.js';

import { api, rgbRandom } from '../../domain';

import ShakePanel from '../ShakePanel/ShakePanel';

function ShakeWindow(Component) {
  return class ShakeWindow extends React.Component {
    static propTypes = {
      pallete: React.PropTypes.shape({}).isRequired,
      sequence: React.PropTypes.arrayOf(React.PropTypes.string).isRequired,
    }

    constructor(props) {
      super(props);

      this.shakeInstance = new ShakeJS({
        threshold: 15,
        timeout: 1500,
      })

      this.state = {
        pallete: {...this.props.pallete},
        sequence: [...this.props.sequence],
        showShakePanel: false,
      };
    }

    componentDidMount() {
      this.shakeInstance.start();
      window.addEventListener('shake', this.onShakeWindow, false);
    }

    componentWillUnmount() {
      this.shakeInstance.stop();
      window.removeEventListener('shake', this.onShakeWindow, false);
    }

    onShakeWindow = (event) => {
      const {sequence, pallete, showShakePanel} = this.state;

      Object.keys(pallete).forEach(colorColumn => {
        pallete[colorColumn].r = rgbRandom();
        pallete[colorColumn].g = rgbRandom();
        pallete[colorColumn].b = rgbRandom();
      })

      this.setState({
        pallete,
        sequence,
        showShakePanel: !showShakePanel,
      }, api.save({ sequence, pallete }));
    }

    onClickToggleShakePanel = () => {
      this.setState(state => ({
        showShakePanel: !state.showShakePanel
      }));
    }

    render() {
      const {showShakePanel, pallete, sequence} = this.state;

      return (
        <div>
          <Component
            pallete={pallete}
            sequence={sequence}
          />
          {showShakePanel && (
            <ShakePanel
              onClickToggleShakePanel={this.onClickToggleShakePanel}
            />
          )}
        </div>
      )
    }
  }
}

export default ShakeWindow;
