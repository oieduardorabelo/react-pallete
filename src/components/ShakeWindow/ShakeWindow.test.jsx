import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import ShakeWindow from './ShakeWindow';
import ShakePanel from '../ShakePanel/ShakePanel';

const WrappedComponent = () => <div></div>

it('renders correctly', () => {
  const wrapper = mount(
    React.createElement(
      ShakeWindow(WrappedComponent),
      { pallete: {}, sequence: [] },
    )
  );

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('updates pallete/sequence state correctly', () => {
  const wrapper = mount(
    React.createElement(
      ShakeWindow(WrappedComponent),
      { pallete: {}, sequence: [] },
    )
  );

  wrapper.setState({
    pallete: { color1: {} },
    sequence: ['color1']
  })

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('ShakePanel displays when showShakePanel state updates correctly', () => {
  const wrapper = mount(
    React.createElement(
      ShakeWindow(WrappedComponent),
      { pallete: {}, sequence: [] },
    )
  );

  wrapper.setState({
    showShakePanel: true
  })

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('ShakePanel hides when click on close button correctly', () => {
  const wrapper = mount(
    React.createElement(
      ShakeWindow(WrappedComponent),
      { pallete: {}, sequence: [] },
    )
  );

  wrapper.setState({
    showShakePanel: true
  })

  expect(toJson(wrapper)).toMatchSnapshot();

  wrapper.find(ShakePanel).find('button').simulate('click')

  expect(toJson(wrapper)).toMatchSnapshot();
});
