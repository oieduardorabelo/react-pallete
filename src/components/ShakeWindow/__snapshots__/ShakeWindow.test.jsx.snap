exports[`test ShakePanel displays when showShakePanel state updates correctly 1`] = `
<ShakeWindow
  pallete={Object {}}
  sequence={Array []}>
  <div>
    <WrappedComponent
      pallete={Object {}}
      sequence={Array []}>
      <div />
    </WrappedComponent>
    <ShakePanel
      onClickToggleShakePanel={[Function]}>
      <div>
        <div
          onClick={[Function]}
          style={
            Object {
              "animation": "fadeIn 160ms linear",
              "backgroundColor": "rgba(0, 0, 0, 0.5)",
              "bottom": 0,
              "display": "flex",
              "left": 0,
              "position": "fixed",
              "right": 0,
              "top": 0,
              "zIndex": 1,
            }
          } />
        <div
          onClick={[Function]}
          style={
            Object {
              "animation": "slideIn 100ms linear",
              "backgroundColor": "#FFFFFF",
              "bottom": 0,
              "display": "flex",
              "flexDirection": "column",
              "left": 0,
              "padding": 20,
              "position": "fixed",
              "right": 0,
              "zIndex": 2,
            }
          }>
          <button
            onClick={[Function]}
            style={
              Object {
                "alignSelf": "flex-end",
              }
            }
            type="button">
            Close
          </button>
          <span>
            Shake it off!
          </span>
          <p>
            When you shake your phone, we set random colors to your pallete!
          </p>
        </div>
      </div>
    </ShakePanel>
  </div>
</ShakeWindow>
`;

exports[`test ShakePanel hides when click on close button correctly 1`] = `
<ShakeWindow
  pallete={Object {}}
  sequence={Array []}>
  <div>
    <WrappedComponent
      pallete={Object {}}
      sequence={Array []}>
      <div />
    </WrappedComponent>
    <ShakePanel
      onClickToggleShakePanel={[Function]}>
      <div>
        <div
          onClick={[Function]}
          style={
            Object {
              "animation": "fadeIn 160ms linear",
              "backgroundColor": "rgba(0, 0, 0, 0.5)",
              "bottom": 0,
              "display": "flex",
              "left": 0,
              "position": "fixed",
              "right": 0,
              "top": 0,
              "zIndex": 1,
            }
          } />
        <div
          onClick={[Function]}
          style={
            Object {
              "animation": "slideIn 100ms linear",
              "backgroundColor": "#FFFFFF",
              "bottom": 0,
              "display": "flex",
              "flexDirection": "column",
              "left": 0,
              "padding": 20,
              "position": "fixed",
              "right": 0,
              "zIndex": 2,
            }
          }>
          <button
            onClick={[Function]}
            style={
              Object {
                "alignSelf": "flex-end",
              }
            }
            type="button">
            Close
          </button>
          <span>
            Shake it off!
          </span>
          <p>
            When you shake your phone, we set random colors to your pallete!
          </p>
        </div>
      </div>
    </ShakePanel>
  </div>
</ShakeWindow>
`;

exports[`test ShakePanel hides when click on close button correctly 2`] = `
<ShakeWindow
  pallete={Object {}}
  sequence={Array []}>
  <div>
    <WrappedComponent
      pallete={Object {}}
      sequence={Array []}>
      <div />
    </WrappedComponent>
  </div>
</ShakeWindow>
`;

exports[`test renders correctly 1`] = `
<ShakeWindow
  pallete={Object {}}
  sequence={Array []}>
  <div>
    <WrappedComponent
      pallete={Object {}}
      sequence={Array []}>
      <div />
    </WrappedComponent>
  </div>
</ShakeWindow>
`;

exports[`test updates pallete/sequence state correctly 1`] = `
<ShakeWindow
  pallete={Object {}}
  sequence={Array []}>
  <div>
    <WrappedComponent
      pallete={
        Object {
          "color1": Object {},
        }
      }
      sequence={
        Array [
          "color1",
        ]
      }>
      <div />
    </WrappedComponent>
  </div>
</ShakeWindow>
`;
